pkg/container/immset.go

```diff
@@ -68,16 +76,50 @@ func (s ImmSet[T]) Delete(xs ...T) ImmSet[T] {
 }
 
 func (s ImmSet[T]) Union(s2 ImmSet[T]) ImmSet[T] {
-	result := make([]T, len(s.xs)+len(s2.xs))
-	copy(result, s.xs)
-	copy(result[len(s.xs):], s2.xs)
-	slices.SortFunc(s.xs, s.cmp)
-	result = slices.CompactFunc(result, s.eq)
+	result := make([]T, 0, len(s.xs)+len(s2.xs))
+	i, j := 0, 0
+	for i < len(s.xs) && j < len(s2.xs) {
+		switch diff := s.cmp(s.xs[i], s2.xs[j]); {
+		case diff < 0:
+			result = append(result, s.xs[i])
+			i++
+		case diff > 0:
+			result = append(result, s2.xs[j])
+			j++
+		default:
+			result = append(result, s.xs[i])
+			i++
+			j++
+		}
+	}
+	for ; i < len(s.xs); i++ {
+		result = append(result, s.xs[i])
+	}
+	for ; j < len(s2.xs); j++ {
+		result = append(result, s2.xs[j])
+	}
 	return ImmSet[T]{result, s.cmp, s.eq}
 }
 
 func (s ImmSet[T]) Difference(s2 ImmSet[T]) ImmSet[T] {
-	return s.Delete(s2.xs...)
+	result := make([]T, 0, len(s.xs))
+	i, j := 0, 0
+	for i < len(s.xs) && j < len(s2.xs) {
```

Comment:
I usually try to avoid indexes and instead update the slice:
```
xs1 := s.xs
xs2 := s2.xs
for len(xs1) > 0 && len(xs2) > 0 {
  switch diff := s.cmp(xs[0], xs2[0]) {
    case diff < 0:
      result = append(result, xs1[0])
      xs1 = xs[1:]
    ...
```

Not sure if this results in worse assembly though.

Reply:
Nice idea! I had this feeling that I was doing C++ in Golang here, and your approach appears way more idiomatic. I'm not sure about the assembly, but I see no obvious impact on the benchmarks, so I'm following the suggestion.